### ISSUE TITLE
Facing python related issues while using sshuttle

### ISSUE THREAD

**Issue by alishaIBM** [sshuttle/sshuttle#673]
I am using a CentOS Linux where I have sshuttle installed.

Installed sshuttle using foll. 
```
git clone https://github.com/sshuttle/sshuttle.git
cd sshuttle
sudo ./setup.py install
```
```
# python --version
Python 3.6.8

# sshuttle --version
1.0.6.dev83+gd3fae53
```

Error seen : 
```
# sudo sshuttle -r root@9.114.xxx.xx 192.168.88.0/24
  File "/usr/share/sshuttle/main.py", line 79
    print version.TAG
                ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(version.TAG)?
```
also tried passing python path : 
```
# sudo sshuttle --python=/usr/bin/python3 -r root@9.114.xxx.xx 192.168.88.0/24
  File "/usr/share/sshuttle/main.py", line 79
    print version.TAG
                ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(version.TAG)?
```

**Comment by brianmay** [comment]
What is `/usr/share/sshuttle/main.py` on your system? Maybe a really old version of sshuttle?

**Comment by alishaIBM** [comment]
Hi @brianmay , 
pasted the output of `/usr/share/sshuttle/main.py` from my system.
```
# cat /usr/share/sshuttle/main.py
import sys, os, re
import helpers, options, client, server, firewall, hostwatch
import compat.ssubprocess as ssubprocess
from helpers import *


# list of:
# 1.2.3.4/5 or just 1.2.3.4
def parse_subnets(subnets_str):
    subnets = []
    for s in subnets_str:
        m = re.match(r'(\d+)(?:\.(\d+)\.(\d+)\.(\d+))?(?:/(\d+))?$', s)
        if not m:
            raise Fatal('%r is not a valid IP subnet format' % s)
        (a,b,c,d,width) = m.groups()
        (a,b,c,d) = (int(a or 0), int(b or 0), int(c or 0), int(d or 0))
        if width == None:
            width = 32
        else:
            width = int(width)
        if a > 255 or b > 255 or c > 255 or d > 255:
            raise Fatal('%d.%d.%d.%d has numbers > 255' % (a,b,c,d))
        if width > 32:
            raise Fatal('*/%d is greater than the maximum of 32' % width)
        subnets.append(('%d.%d.%d.%d' % (a,b,c,d), width))
    return subnets


# 1.2.3.4:567 or just 1.2.3.4 or just 567
def parse_ipport(s):
    s = str(s)
    m = re.match(r'(?:(\d+)\.(\d+)\.(\d+)\.(\d+))?(?::)?(?:(\d+))?$', s)
    if not m:
        raise Fatal('%r is not a valid IP:port format' % s)
    (a,b,c,d,port) = m.groups()
    (a,b,c,d,port) = (int(a or 0), int(b or 0), int(c or 0), int(d or 0),
                      int(port or 0))
    if a > 255 or b > 255 or c > 255 or d > 255:
        raise Fatal('%d.%d.%d.%d has numbers > 255' % (a,b,c,d))
    if port > 65535:
        raise Fatal('*:%d is greater than the maximum of 65535' % port)
    if a == None:
        a = b = c = d = 0
    return ('%d.%d.%d.%d' % (a,b,c,d), port)


optspec = """
sshuttle [-l [ip:]port] [-r [username@]sshserver[:port]] <subnets...>
sshuttle --server
sshuttle --firewall <port> <subnets...>
sshuttle --hostwatch
--
l,listen=  transproxy to this ip address and port number [127.0.0.1:0]
H,auto-hosts scan for remote hostnames and update local /etc/hosts
N,auto-nets  automatically determine subnets to route
dns        capture local DNS requests and forward to the remote DNS server
python=    path to python interpreter on the remote server
r,remote=  ssh hostname (and optional username) of remote sshuttle server
x,exclude= exclude this subnet (can be used more than once)
exclude-from=  exclude the subnets in a file (whitespace separated)
v,verbose  increase debug message verbosity
e,ssh-cmd= the command to use to connect to the remote [ssh]
seed-hosts= with -H, use these hostnames for initial scan (comma-separated)
no-latency-control  sacrifice latency to improve bandwidth benchmarks
wrap=      restart counting channel numbers after this number (for testing)
D,daemon   run in the background as a daemon
V,version  print sshuttle's version number
syslog     send log messages to syslog (default if you use --daemon)
pidfile=   pidfile name (only if using --daemon) [./sshuttle.pid]
server     (internal use only)
firewall   (internal use only)
hostwatch  (internal use only)
"""
o = options.Options(optspec)
(opt, flags, extra) = o.parse(sys.argv[2:])

if opt.version:
    import version
    print version.TAG
    sys.exit(0)
if opt.daemon:
    opt.syslog = 1
if opt.wrap:
    import ssnet
    ssnet.MAX_CHANNEL = int(opt.wrap)
helpers.verbose = opt.verbose

try:
    if opt.server:
        if len(extra) != 0:
            o.fatal('no arguments expected')
        server.latency_control = opt.latency_control
        sys.exit(server.main())
    elif opt.firewall:
        if len(extra) != 2:
            o.fatal('exactly two arguments expected')
        sys.exit(firewall.main(int(extra[0]), int(extra[1]), opt.syslog))
    elif opt.hostwatch:
        sys.exit(hostwatch.hw_main(extra))
    else:
        if len(extra) < 1 and not opt.auto_nets:
            o.fatal('at least one subnet (or -N) expected')
        includes = extra
        excludes = ['127.0.0.0/8']
        for k,v in flags:
            if k in ('-x','--exclude'):
                excludes.append(v)
            if k in ('-X', '--exclude-from'):
                excludes += open(v).read().split()
        remotename = opt.remote
        if remotename == '' or remotename == '-':
            remotename = None
        if opt.seed_hosts and not opt.auto_hosts:
            o.fatal('--seed-hosts only works if you also use -H')
        if opt.seed_hosts:
            sh = re.split(r'[\s,]+', (opt.seed_hosts or "").strip())
        elif opt.auto_hosts:
            sh = []
        else:
            sh = None
        sys.exit(client.main(parse_ipport(opt.listen or '0.0.0.0:0'),
                             opt.ssh_cmd,
                             remotename,
                             opt.python,
                             opt.latency_control,
                             opt.dns,
                             sh,
                             opt.auto_nets,
                             parse_subnets(includes),
                             parse_subnets(excludes),
                             opt.syslog, opt.daemon, opt.pidfile))
except FatalNeedsReboot, e:
    log('You must reboot before using sshuttle.\n')
    sys.exit(EXITCODE_NEEDS_REBOOT)
except Fatal, e:
    log('fatal: %s\n' % e)
    sys.exit(99)
except KeyboardInterrupt:
    log('\n')
    log('Keyboard interrupt: exiting.\n')
    sys.exit(1)
```

I had cloned the project freshly to install sshuttle : 
```
git clone https://github.com/sshuttle/sshuttle.git
cd sshuttle
sudo ./setup.py install
```
```
# sshuttle --version
1.0.6.dev83+gd3fae53
```

**Comment by brianmay** [comment]
I think you might have old files lying around on you system. Recent versions of sshuttle don't come with anything like a that file.

Maybe move/delete your entire `/usr/share/sshuttle` directory?

**Comment by alishaIBM** [comment]
These are the locations : 
```
# which sshuttle
/usr/local/bin/sshuttle
```
```
# whereis sshuttle
sshuttle: /usr/bin/sshuttle /usr/local/bin/sshuttle /usr/share/sshuttle /usr/share/man/man8/sshuttle.8.gz
```
FYI : @brianmay

**Comment by alishaIBM** [comment]
Tried foll. : 

- removed /usr/share/sshuttle
- re-tried with : 
```
# sshuttle --version
1.0.6.dev85+ge59cc39

# which sshuttle
/usr/local/bin/sshuttle

# whereis sshuttle
sshuttle: /usr/local/bin/sshuttle

# python --version
Python 3.6.8

# which python
/usr/bin/python
```
Error : 
```
# sshuttle -r root@9.114.181.66 192.168.88.0/24
Traceback (most recent call last):
  File "/usr/local/bin/sshuttle", line 11, in <module>
    load_entry_point('sshuttle==1.0.6.dev85+ge59cc39', 'console_scripts', 'sshuttle')()
  File "/usr/local/lib/python3.6/site-packages/sshuttle-1.0.6.dev85+ge59cc39-py3.6.egg/sshuttle/cmdline.py", line 119, in main
    opt.tmark)
  File "/usr/local/lib/python3.6/site-packages/sshuttle-1.0.6.dev85+ge59cc39-py3.6.egg/sshuttle/client.py", line 919, in main
    raise e
  File "/usr/local/lib/python3.6/site-packages/sshuttle-1.0.6.dev85+ge59cc39-py3.6.egg/sshuttle/client.py", line 908, in main
    tcp_listener.bind(lv6, lv4)
  File "/usr/local/lib/python3.6/site-packages/sshuttle-1.0.6.dev85+ge59cc39-py3.6.egg/sshuttle/client.py", line 179, in bind
    self.v6.bind(address_v6)
OSError: [Errno 99] Cannot assign requested address
```

**Comment by alishaIBM** [comment]
What should ideally be the latest version of sshuttle to be installed on x86_64 ?

**Comment by brianmay** [comment]
The version you are using should be fine.

But for some reason not working on your system. Looks like it is trying to do an IPv6 bind. Which is unexpected here, because you haven't specified a method that supports ipv6.

Maybe try `--disable-ipv6`?

**Comment by alishaIBM** [comment]
yes , the above option helped.

```
sshuttle --disable-ipv6 -r root@9.114.xxx.xx 192.168.88.0/24
root@9.114.xxx.xx's password:
c : Connected to server.
```

Thanks @brianmay !

**Comment by skuhl** [comment]
> But for some reason not working on your system. Looks like it is trying to do an IPv6 bind. Which is unexpected here, because you haven't specified a method that supports ipv6.

We added IPv6 support the "nat" method (i.e., iptables): #646

I've used the latest code on a variety of machines supporting IPv4+IPv6 and just IPv4 and haven't had trouble. But, it is possible something got missed. Seems like there is room for improvement in the code here.

**Comment by brianmay** [comment]
@skuhl Thanks, I had forgotten about that change. We need to be able to understand the failure before we can fix it. Normally "Cannot assign requested address" would mean that the address is in use, which seems strange. Unless maybe sshuttle hadn't exit cleanly or something.

@alishaIBM Are you able to run the following command and paste the output here? Thanks

```
sshuttle -vvv -r root@9.114.xxx.xx 192.168.88.0/24
```

**Comment by alishaIBM** [comment]
yes !
```
# sshuttle -vvv -r root@9.114.xxx.xx 192.168.88.0/24
Starting sshuttle proxy (version 1.0.6.dev85+ge59cc39).
c : which() found 'sudo' at /usr/bin/sudo
fw: Starting firewall with Python version 3.6.8
fw: Selecting a method automatically...
fw: which() found 'iptables' at /usr/sbin/iptables
fw: Method 'nat' was automatically selected.
fw: which() found 'iptables' at /usr/sbin/iptables
fw: ready method name nat.
c : IPv6 enabled: Using default IPv6 listen address ::1
c : Method: nat
c : IPv4: on
c : IPv6: on
c : UDP : off (not available with nat method)
c : DNS : off (available)
c : User: off (available)
c : Subnets to forward through remote host (type, IP, cidr mask width, startPort, endPort):
c :   (<AddressFamily.AF_INET: 2>, '192.168.88.0', 24, 0, 0)
c : Subnets to exclude from forwarding:
c :   (<AddressFamily.AF_INET: 2>, '127.0.0.1', 32, 0, 0)
c :   (<AddressFamily.AF_INET6: 10>, '::1', 128, 0, 0)
c : Trying to bind redirector on port 12300
Traceback (most recent call last):
  File "/usr/local/bin/sshuttle", line 11, in <module>
    load_entry_point('sshuttle==1.0.6.dev85+ge59cc39', 'console_scripts', 'sshuttle')()
  File "/usr/local/lib/python3.6/site-packages/sshuttle-1.0.6.dev85+ge59cc39-py3.6.egg/sshuttle/cmdline.py", line 119, in main
    opt.tmark)
  File "/usr/local/lib/python3.6/site-packages/sshuttle-1.0.6.dev85+ge59cc39-py3.6.egg/sshuttle/client.py", line 919, in main
    raise e
  File "/usr/local/lib/python3.6/site-packages/sshuttle-1.0.6.dev85+ge59cc39-py3.6.egg/sshuttle/client.py", line 908, in main
    tcp_listener.bind(lv6, lv4)
  File "/usr/local/lib/python3.6/site-packages/sshuttle-1.0.6.dev85+ge59cc39-py3.6.egg/sshuttle/client.py", line 179, in bind
    self.v6.bind(address_v6)
OSError: [Errno 99] Cannot assign requested address
```

**Comment by skuhl** [comment]
I can reproduce this by disabling IPv6 entirely. I believe on many machines where IPv6 isn't present, it is still supported locally. Running these commands as root should entirely disable IPv6 until reboot (or changing the setting back to 0). When I do so, I get the same error message. It probably applies to all of the different methods that support IPv6, but I just tested iptables.

```
echo 1 > /proc/sys/net/ipv6/conf/all/disable_ipv6
echo 1 > /proc/sys/net/ipv6/conf/default/disable_ipv6
```

**Comment by brianmay** [comment]
Wonder if "Cannot assign requested address" is some sort of generic error because we are trying to bind on ipv6 address, and there is no ipv6.

Fixing this might be tricky. The naive approach would be to ignore "Cannot assign requested address" errors when binding. But if that happens, we need to make sure we don't set-up the ipv6 firewall  to point to the non-existent bind, as that might cause problems also.

https://github.com/sshuttle/sshuttle/blob/master/sshuttle/client.py#L179

An added complication is the code where this is currently failing is dodgy code for searching for a free port to bind on.

https://github.com/sshuttle/sshuttle/blob/master/sshuttle/client.py#L878-L919

Which is then completely duplicated to find a free port to bind on for DNS. So both of these loops need to do something sensible also. Maybe even extract the common stuff into a function called "get_free_port" or something.

https://github.com/sshuttle/sshuttle/blob/master/sshuttle/client.py#L933-L964

**Comment by skuhl** [comment]
An easy solution: When this error occurs, print an message saying that one cause of the error is running sshuttle with IPv6 disabled and tell the user that rerunning with --disable-ipv6 may resolve the problem.

I'm sure there are also more elegant solutions. There are plenty of opportunities for code cleanup as well.

**Comment by brianmay** [comment]
Closed in #728.
